ReviewModel lemmatized_text and sentence_vector features, added validate method

alexandr-z opened the pull request:
- added lemmatized_text and sentence_vector features
- added class_weight='balanced' to SGDClassifier
- added validate method

alexandr-z committed:
ReviewModel lemmatized_text and sentence_vector features, added validate method

bdewilde commented on colandr/lib/extractors/review_model.py:
```diff
@@ -100,15 +121,18 @@ def train(self, training_data: list[TrainingData], min_samples: int = 40) -> boo
         sgd_classifier = SGDClassifier(
             loss='log_loss', random_state=42, early_stopping=True,
-            n_iter_no_change=10, alpha=5e-4
+            n_iter_no_change=10, alpha=5e-4, class_weight='balanced'
```
👍

bdewilde commented on colandr/lib/extractors/review_model.py:
```diff
@@ -228,85 +285,100 @@ def _prepare_data_for_training(
             Tuple containing the feature DataFrame (X) and target array (y).
+        """
+        x, targets = self._create_sentence_features(training_data)
+        self.label_binarizer = MultiLabelBinarizer()
```
this line surprised me -- I was trying to figure out where `label_binarizer` got set, and didn't expect it to be in the method for preparing training data. would it make more sense somewhere else?

alexandr-z replied:
Moved to the train method

bdewilde reviewed (approved):
I didn't see anything obviously wrong. However, since many of your functions changed outputs, I'd have expected some of your unit tests to need updating -- is that not the case? 

The code is a bit hard to follow, at least to my eyes. Would it be possible to add a bit of documentation to, say, the `ReviewModel` class docstring describing what it's doing in words, and maybe giving a quick usage example?

alexandr-z committed:
Moved MultiLabelBinarizer to the train method

alexandr-z committed:
Added extended class docstring to ReviewModel

alexandr-z commented:
> I didn't see anything obviously wrong. However, since many of your functions changed outputs, I'd have expected some of your unit tests to need updating -- is that not the case?
> 
> The code is a bit hard to follow, at least to my eyes. Would it be possible to add a bit of documentation to, say, the `ReviewModel` class docstring describing what it's doing in words, and maybe giving a quick usage example?

After the initial commit the tests remained unchanged because `_prepare_data_for_training` was split but remained with the same signature. When I removed this method which became just a proxy the corresponding tests were updated accordingly.

Added extended class docstring.
